File: matlab-package/brainflow/MLModel.m
classdef MLModel
    % MLModel for inference
    properties
        input_json
        input_params
    end

    methods(Static)
        
        function lib_name = load_lib()
            if ispc
                if not(libisloaded('MLModule'))
                    loadlibrary('MLModule.dll', 'ml_module.h');
                end
                lib_name = 'MLModule';
            elseif ismac
                if not(libisloaded('libMLModule'))
                    loadlibrary('libMLModule.dylib', 'ml_module.h');
                end
                lib_name = 'libMLModule';
            elseif isunix
                if not(libisloaded('libMLModule'))
                    loadlibrary('libMLModule.so', 'ml_module.h');
                end
                lib_name = 'libMLModule';
            else
                error('OS not supported!')
            end
        end
        
        function check_ec(ec, task_name)
            if(ec ~= int32(BrainFlowExitCodes.STATUS_OK))
                error('Non zero ec: %d, for task: %s', ec, task_name)
            end
        end
        
        function release_all()
            % release all sessions
            task_name = 'release_all';
            lib_name = MLModel.load_lib();
            exit_code = calllib(lib_name, task_name);
            MLModel.check_ec(exit_code, task_name);
        end

        function log_message(log_level, message)
            % write message to ML logger
            task_name = 'log_message_ml_module';
            lib_name = MLModel.load_lib();
            exit_code = calllib(lib_name, task_name, log_level, message);
            MLModel.check_ec(exit_code, task_name);
        end
        
        function set_log_level(log_level)
            % set log level for MLModel
            task_name = 'set_log_level_ml_module';
            lib_name = MLModel.load_lib();
            exit_code = calllib(lib_name, task_name, log_level);
            MLModel.check_ec(exit_code, task_name);
        end

        function set_log_file(log_file)
            % set log file for MLModel
            task_name = 'set_log_file_ml_module';
            lib_name = MLModel.load_lib();
            exit_code = calllib(lib_name, task_name, log_file);
            MLModel.check_ec(exit_code, task_name);
        end

        function enable_ml_logger()
            % enable logger with level INFO
            MLModel.set_log_level(int32(2))
        end

        function enable_dev_ml_logger()
            % enable logger with level TRACE
            MLModel.set_log_level(int32(0))
        end

        function disable_ml_logger()
            % disable logger
            MLModel.set_log_level(int32(6))
        end
        
        function version = get_version()
            % get version
            task_name = 'get_version_ml_module';
            lib_name = MLModel.load_lib();
            % no way to understand how it works in matlab, used this link
            % https://nl.mathworks.com/matlabcentral/answers/131446-what-data-type-do-i-need-to-calllib-with-pointer-argument-char%
            [exit_code, version] = calllib(lib_name, task_name, blanks(64), 64, 64);
            MLModel.check_ec(exit_code, task_name);
        end
   
    end

    methods

        function obj = MLModel(params)
            obj.input_json = params.to_json();
            obj.input_params = params;
        end

        function prepare(obj)
            % prepare model
            task_name = 'prepare';
            lib_name = MLModel.load_lib();
            exit_code = calllib(lib_name, task_name, obj.input_json);
            MLModel.check_ec(exit_code, task_name);
        end
        
        function release(obj)
            % release model
            task_name = 'release';
            lib_name = MLModel.load_lib();
            exit_code = calllib(lib_name, task_name, obj.input_json);
            MLModel.check_ec(exit_code, task_name);
        end

        function score = predict(obj, input_data)
            % perform inference for input data
            task_name = 'predict';
            lib_name = MLModel.load_lib();
            score_temp = libpointer('doublePtr', obj.input_params.max_array_size);
            len = libpointer('int32Ptr', 0);
            input_data_temp = libpointer('doublePtr', input_data);
            exit_code = calllib(lib_name, task_name, input_data_temp, size(input_data, 2), score_temp, len, obj.input_json);
            MLModel.check_ec(exit_code, task_name);
            score = score_temp.Value(1,1:len.Value);
        end
        
    end
    
end

File: java-package/brainflow/src/main/java/brainflow/MLModel.java
package brainflow;

import java.io.File;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;

import org.apache.commons.lang3.SystemUtils;

import com.sun.jna.Library;
import com.sun.jna.Native;

public class MLModel
{
    private interface DllInterface extends Library
    {

        int set_log_level_ml_module (int log_level);

        int set_log_file_ml_module (String log_file);

        int prepare (String params);

        int release (String params);

        int predict (double[] data, int data_len, double[] output, String params);
    }

    private static DllInterface instance;

    static
    {
        // SystemUtils doesnt have smth like IS_OS_ANDROID, need to check using
        // properties
        boolean is_os_android = "The Android Project".equals (System.getProperty ("java.specification.vendor"));

        String lib_name = "libMLModule.so";
        if (SystemUtils.IS_OS_WINDOWS)
        {
            lib_name = "MLModule.dll";

        } else if (SystemUtils.IS_OS_MAC)
        {
            lib_name = "libMLModule.dylib";
        }

        if (is_os_android)
        {
            // for android you need to put these files manually to jniLibs folder, unpacking
            // doesnt work
            lib_name = "MLModule"; // no lib prefix and no extension for android
        } else
        {
            // need to extract libraries from jar
            unpack_from_jar (lib_name);
            unpack_from_jar ("brainflow_svm.model");
        }

        instance = Native.loadLibrary (lib_name, DllInterface.class);
    }

    private static Path unpack_from_jar (String lib_name)
    {
        try
        {
            File file = new File (lib_name);
            if (file.exists ())
                file.delete ();
            InputStream link = (BoardShim.class.getResourceAsStream (lib_name));
            Files.copy (link, file.getAbsoluteFile ().toPath ());
            return file.getAbsoluteFile ().toPath ();
        } catch (Exception io)
        {
            System.err.println ("file: " + lib_name + " is not found in jar file");
            return null;
        }
    }

    private String input_params;

    /**
     * Create MLModel object
     */
    public MLModel (BrainFlowModelParams params)
    {
        input_params = params.to_json ();
    }

    /**
     * enable ML logger with level INFO
     */
    public static void enable_ml_logger () throws BrainFlowError
    {
        set_log_level (2);
    }

    /**
     * enable ML logger with level TRACE
     */
    public static void enable_dev_ml_logger () throws BrainFlowError
    {
        set_log_level (0);
    }

    /**
     * disable BrainFlow logger
     */
    public static void disable_ml_logger () throws BrainFlowError
    {
        set_log_level (6);
    }

    /**
     * redirect logger from stderr to a file
     */
    public static void set_log_file (String log_file) throws BrainFlowError
    {
        int ec = instance.set_log_file_ml_module (log_file);
        if (ec != ExitCode.STATUS_OK.get_code ())
        {
            throw new BrainFlowError ("Error in set_log_file", ec);
        }
    }

    /**
     * set log level
     */
    private static void set_log_level (int log_level) throws BrainFlowError
    {
        int ec = instance.set_log_level_ml_module (log_level);
        if (ec != ExitCode.STATUS_OK.get_code ())
        {
            throw new BrainFlowError ("Error in set_log_level", ec);
        }
    }

    /**
     * Prepare classifier
     * 
     * @throws BrainFlowError
     */
    public void prepare () throws BrainFlowError
    {
        int ec = instance.prepare (input_params);
        if (ec != ExitCode.STATUS_OK.get_code ())
        {
            throw new BrainFlowError ("Error in prepare", ec);
        }
    }

    /**
     * Release classifier
     * 
     * @throws BrainFlowError
     */
    public void release () throws BrainFlowError
    {
        int ec = instance.release (input_params);
        if (ec != ExitCode.STATUS_OK.get_code ())
        {
            throw new BrainFlowError ("Error in release", ec);
        }
    }

    /**
     * Get score of classifier
     * 
     * @throws BrainFlowError
     */
    public double predict (double[] data) throws BrainFlowError
    {
        double[] val = new double[1];
        int ec = instance.predict (data, data.length, val, input_params);
        if (ec != ExitCode.STATUS_OK.get_code ())
        {
            throw new BrainFlowError ("Error in predict", ec);
        }
        return val[0];
    }
}
